src/main/java/org/prebid/server/bidder/huaweiads/HuaweiAdsBidder.java

```diff
@@ -0,0 +1,1311 @@
+
+            return extUser.getConsent();
+        }
+        return null;
+    }
+
+    private MultiMap getHeaders(ExtImpHuaweiAds huaweiAdsImpExt, BidRequest request) {
+        final MultiMap headers = HttpUtil.headers();
+
+        if (huaweiAdsImpExt == null) {
+            return headers;
+        }
+
+        final String authorization = getDigestAuthorization(huaweiAdsImpExt);
+
+        headers.set(HttpUtil.AUTHORIZATION_HEADER, authorization);
+
+        Optional.ofNullable(request.getDevice())
+                .map(com.iab.openrtb.request.Device::getUa)
+                .filter(StringUtils::isNotEmpty)
+                .ifPresent(ua -> headers.set(HttpUtil.USER_AGENT_HEADER, ua));
+
+        return headers;
+    }
+
+    private String getDigestAuthorization(ExtImpHuaweiAds huaweiAdsImpExt) {
+        final String nonce = String.valueOf(System.currentTimeMillis());
+
+        final String apiKey = huaweiAdsImpExt.getPublisherId() + ":ppsadx/getResult:" + huaweiAdsImpExt.getSignKey();
+
+        final String data = nonce + ":POST:/ppsadx/getResult";
+        final String hmacSha256 = computeHmacSha256(data, apiKey);
+
+        return "Digest "
+                + "username=" + huaweiAdsImpExt.getPublisherId()
+                + ",realm=ppsadx/getResult,"
+                + "nonce=" + nonce
+                + ",response=" + hmacSha256 + ","
+                + "algorithm=HmacSHA256,usertype=1,keyid=" + huaweiAdsImpExt.getKeyId();
+    }
+
+    private String computeHmacSha256(String message, String signKey) {
+        final String algorithm = "HmacSHA256";
+
+        try {
+            final SecretKeySpec secretKeySpec = new SecretKeySpec(signKey.getBytes(StandardCharsets.UTF_8), algorithm);
+            final Mac mac = Mac.getInstance(algorithm);
+            mac.init(secretKeySpec);
+
+            final byte[] hmac = mac.doFinal(message.getBytes(StandardCharsets.UTF_8));
+            return bytesToHex(hmac);
+        } catch (NoSuchAlgorithmException | InvalidKeyException e) {
+            throw new PreBidException(e.getMessage());
+        }
+    }
+
+    private String bytesToHex(byte[] bytes) {
+        final StringBuilder hex = new StringBuilder();
+        for (byte b : bytes) {
+            hex.append(String.format("%02x", b));
+        }
+        return hex.toString();
+    }
+
+    private String buildEndpoint(String countryCode) {
+        if (countryCode == null || countryCode.length() > 2 || "1".equals(closeSiteSelectionByCountry)) {
+            return endpointUrl;
+        }
+
+        // choose site
+        if (ChineseSiteCountryCode.isContainsByName(countryCode)) {
+            return chineseEndpoint;
+        } else if (RussianSiteCountryCode.isContainsByName(countryCode)) {
+            return russianEndpoint;
+        } else if (EuropeanSiteCountryCode.isContainsByName(countryCode)) {
+            return europeanEndpoint;
+        } else {
+            return asianEndpoint;
+        }
+    }
```

Comment:
What about to extract this logic + `ChineseSiteCountryCode` + `RussianSiteCountryCode` + `EuropeanSiteCountryCode` in separated configurable object that decides what endpoint must be provided for certain country code?

```
SomeResolver(endpoint1, endpoint2, ...)  {
 static Set<String> europeanCodes = Set.of("AA", "AB", ...)
 
 public String for(countryCode) {
    if (...) return  endpointUrl;
    
    final String normalizedCountryCode = countryCode.toUpperCase();
    if (normalizedCountryCode == "CN") return chineseEndpoint;
    if (normalizedCountryCode == "RU") return russianEndpoint;
    if (europeanCodes.contains(normalizedCountryCode)) return europeanEndpoint;

    return asianEndpoint;
 }
}
```